// react-ckeditor-classic/lib/editor-config.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _autoformat = _interopRequireDefault(require("@ckeditor/ckeditor5-autoformat/src/autoformat"));

var _alignment = _interopRequireDefault(require("@ckeditor/ckeditor5-alignment/src/alignment"));

var _essentials = _interopRequireDefault(require("@ckeditor/ckeditor5-essentials/src/essentials"));

var _paragraph = _interopRequireDefault(require("@ckeditor/ckeditor5-paragraph/src/paragraph"));

var _indent = _interopRequireDefault(require("@ckeditor/ckeditor5-indent/src/indent"));

var _bold = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/bold"));

var _italic = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/italic"));

var _underline = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/underline"));

var _strikethrough = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/strikethrough"));

var _code = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/code"));

var _subscript = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/subscript"));

var _superscript = _interopRequireDefault(require("@ckeditor/ckeditor5-basic-styles/src/superscript"));

var _link = _interopRequireDefault(require("@ckeditor/ckeditor5-link/src/link"));

var _font = _interopRequireDefault(require("@ckeditor/ckeditor5-font/src/font"));

var _table = _interopRequireDefault(require("@ckeditor/ckeditor5-table/src/table"));

var _tabletoolbar = _interopRequireDefault(require("@ckeditor/ckeditor5-table/src/tabletoolbar"));

var _list = _interopRequireDefault(require("@ckeditor/ckeditor5-list/src/list"));

var _blockquote = _interopRequireDefault(require("@ckeditor/ckeditor5-block-quote/src/blockquote"));

var _heading = _interopRequireDefault(require("@ckeditor/ckeditor5-heading/src/heading"));

var _image = _interopRequireDefault(require("@ckeditor/ckeditor5-image/src/image"));

var _imagecaption = _interopRequireDefault(require("@ckeditor/ckeditor5-image/src/imagecaption"));

var _imagestyle = _interopRequireDefault(require("@ckeditor/ckeditor5-image/src/imagestyle"));

var _imagetoolbar = _interopRequireDefault(require("@ckeditor/ckeditor5-image/src/imagetoolbar"));

var _imageupload = _interopRequireDefault(require("@ckeditor/ckeditor5-image/src/imageupload"));

var _removeformat = _interopRequireDefault(require("@ckeditor/ckeditor5-remove-format/src/removeformat"));

var _UploadAdapterPlugin = _interopRequireDefault(require("./UploadAdapterPlugin"));

// import CKFinderUploadAdapter from '@ckeditor/ckeditor5-adapter-ckfinder/src/uploadadapter';
var editorConfig = {
  plugins: [_essentials["default"], _paragraph["default"], _indent["default"], _bold["default"], _italic["default"], _underline["default"], _strikethrough["default"], _subscript["default"], _superscript["default"], _link["default"], _code["default"], _heading["default"], _blockquote["default"], _list["default"], _table["default"], _tabletoolbar["default"], _image["default"], _imagecaption["default"], _imagestyle["default"], _imagetoolbar["default"], _imageupload["default"], // CKFinderUploadAdapter,
  _UploadAdapterPlugin["default"], _alignment["default"], _autoformat["default"], _font["default"], _removeformat["default"]],
  toolbar: ['heading', '|', 'outdent', 'indent', 'bold', 'italic', 'underline', 'strikethrough', 'subscript', 'superscript', 'fontSize', 'fontFamily', 'fontColor', 'fontBackgroundColor', 'alignment', 'link', 'code', 'bulletedList', 'numberedList', 'insertTable', 'blockQuote', 'undo', 'redo', 'imageUpload', 'removeFormat'],
  heading: {
    options: [{
      model: 'paragraph',
      title: 'Paragraph',
      "class": 'ck-heading_paragraph'
    }, {
      model: 'heading1',
      view: 'h1',
      title: 'Heading 1',
      "class": 'ck-heading_heading1'
    }, {
      model: 'heading2',
      view: 'h2',
      title: 'Heading 2',
      "class": 'ck-heading_heading2'
    }, {
      model: 'heading3',
      view: 'h3',
      title: 'Heading 3',
      "class": 'ck-heading_heading3'
    }]
  },
  fontSize: {
    options: [10, 12, 14, 16, 18, 22, 24, 28, 32, 64]
  },
  fontFamily: {
    options: ['default', 'Arial, Helvetica, sans-serif', 'Courier New, Courier, monospace', 'Georgia, serif', 'Lucida Sans Unicode, Lucida Grande, sans-serif', 'Tahoma, Geneva, sans-serif', 'Times New Roman, Times, serif', 'Trebuchet MS, Helvetica, sans-serif', 'Verdana, Geneva, sans-serif']
  },
  fontColor: {
    columns: 6,
    documentColors: 12
  },
  fontBackgroundColor: {
    columns: 6,
    documentColors: 12
  },
  table: {
    contentToolbar: ['tableColumn', 'tableRow', 'mergeTableCells']
  },
  image: {
    toolbar: ['imageStyle:alignLeft', 'imageStyle:full', 'imageStyle:alignRight', // 'imageStyle:side',
    '|', 'imageTextAlternative'],
    styles: ['full', 'alignLeft', 'alignRight']
  },
  alignment: {
    options: ['left', 'right', 'center', 'justify']
  },
  indentBlock: {
    offset: 1,
    unit: 'em'
  }
};
var _default = editorConfig;
exports["default"] = _default;
